File: haskell/src/validBST.hs
module ValidBST where

-- https://en.wikipedia.org/wiki/Tree_traversal#Pre-order

-- data Tree = Node Tree Tree | Leaf

-- foo = Node (Node Leaf Leaf) Leaf

-- https://dkalemis.wordpress.com/2014/01/23/trees-in-haskell/
data Tree a = Nil | Node a (Tree a) (Tree a)

foo = Node 1 (Node 3 Nil Nil) (Node 2 Nil Nil)

insertElement :: (Ord a, Eq a) => Tree a -> a -> Tree a
insertElement Nil x = Node x Nil Nil
insertElement t@(Node y left right) x
    | x < y = Node y (insertElement left x) right
    | x > y = Node y left (insertElement right x)
    | otherwise = t

-- foldr :: Foldable t => (a -> b -> b) -> b -> t a -> b
instance Foldable Tree where
    foldr :: (a -> b -> b) -> b -> Tree a -> b
    foldr f z Nil = z
    foldr f z (Node a left right) = 
        

File: scala/src/main/scala/ValidBST.scala

object Solution {
  def main(args: Array[String]) = ValidBST.main(args)
}


object ValidBST {
  val sc = new java.util.Scanner (System.in);

  def readTraversal(): List[Int] = {
    val n = sc.nextInt()
    var arr = new Array[Int](n)
    for(i <- 0 to n-1)
      arr(i) = sc.nextInt()

    arr.toList
  }

  // https://www.hackerrank.com/challenges/valid-bst/forum/comments/101479

  trait Tree
  case class Node(x: Int, left: Tree, right: Tree) extends Tree
  case object Empty extends Tree

  def constructNode(head: Int, arr: List[Int]): Tree = {
    val leftArr = arr.filter(_<head)
    val rightArr = arr.filter(_>head)

    val leftTree = leftArr match {
      case List() => Empty
      case x::xs => constructNode(x,xs)
    }
    val rightTree = rightArr match {
      case List() => Empty
      case y::ys => constructNode(y,ys)
    }

    Node(head, leftTree, rightTree)
  }

  def constructTree(arr: List[Int]): Tree = arr match {
    case x::xs => constructNode(x, xs)
    case List() => Empty
  }

  // https://en.wikipedia.org/wiki/Tree_traversal#Post-order
  def postOrderTraversal(t: Tree): List[Int] = t match {
    case Empty => List()
      //case Node(x, leftT, rightT) => postOrderTraversal(leftT):::postOrderTraversal(rightT):::List(x)
    case Node(x, leftT, rightT) =>
      x::postOrderTraversal(leftT):::postOrderTraversal(rightT)
  }


  def main(args: Array[String]) {
    val n = sc.nextInt()
    for(i <- 1 to n) {
      val arr = readTraversal()
      val t = constructTree(arr)
      val arr2 = postOrderTraversal(t)
      if(arr.sameElements(arr2)){
        println("YES")
      } else {
        println("NO")
      }
    }

  }
}
